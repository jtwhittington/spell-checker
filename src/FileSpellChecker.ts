import { readFileSync } from "fs";
import { join } from "path";
import { SpellCheck } from "./SpellCheck";
import { Trie } from "./Trie";
import { SpellingCorrection } from "./types";

export class FileSpellChecker {
  dictionaryTrie: Trie;
  fileToCheck: string;

  constructor(dictionaryFile: string, fileToCheck: string) {
    this.dictionaryTrie = new Trie();
    const contents = readFileSync(join(__dirname, dictionaryFile), "utf-8");

    const wordList = contents.split(/\r?\n/);
    wordList.forEach((word) => this.dictionaryTrie.insert(word));
    this.fileToCheck = fileToCheck;
  }

  checkFile() {
    const misspellings: Array<SpellingCorrection> = [];

    const fileBuffer = readFileSync(join(__dirname, this.fileToCheck), "utf-8");
    const lines = fileBuffer.split(/\r?\n/);
    let prevEndOfSentence = false;
    const sentenceDelimiters = [".", "!", "?"];

    lines.forEach((line, lineNumber) => {
      lineNumber += 1;
      let columnNumber = 0;
      const words = line.split(" ");
      words.forEach((word, wordIndex) => {
        columnNumber += word.length;
        if (word === "") {
          return;
        }

        // if the word is capitalized and previous word was not the end of a sentence, then it's probably a proper noun
        const properNoun =
          word[0] === word[0].toUpperCase() && !prevEndOfSentence;

        // Remove punctuation
        const cleanedWord = word.replace(/\W/g, "");

        // Very naive detection of proper nouns
        const [isCorrect, suggestions] = properNoun
          ? [true, []]
          : new SpellCheck(this.dictionaryTrie, cleanedWord).check();

        prevEndOfSentence = sentenceDelimiters.includes(word[word.length - 1]);

        // get context 2 words before and after
        const context = words
          .slice(Math.max(wordIndex - 2, 0), wordIndex + 3)
          .join(" ");

        if (!isCorrect) {
          misspellings.push({
            word: cleanedWord,
            suggestions,
            context,
            lineNumber,
            columnNumber,
          });
        }
      });
    });
    return misspellings;
  }
}
