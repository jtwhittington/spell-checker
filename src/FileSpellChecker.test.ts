import { FileSpellChecker } from "./FileSpellChecker";

describe("FileSpellChecker", () => {
  it("accepts a dictionary file and file to check and logs any misspelled words", () => {
    const dictionary = "../fixtures/dictionary.txt";
    const fileToCheck = "../fixtures/Small document.txt";

    const checker = new FileSpellChecker(dictionary, fileToCheck);
    expect(checker.checkFile()).toMatchSnapshot();
  });
});
