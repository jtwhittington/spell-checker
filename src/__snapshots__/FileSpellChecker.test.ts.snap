// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileSpellChecker accepts a dictionary file and file to check and logs any misspelled words 1`] = `
[
  {
    "columnNumber": 7,
    "context": "The quik brown fox",
    "lineNumber": 1,
    "suggestions": [
      "quid",
      "quin",
      "quip",
      "quit",
      "quiz",
      "quick",
      "quirk",
      "qua",
      "quack",
      "quad",
      "quads",
      "quaff",
      "quag",
      "quags",
      "quai",
      "quail",
      "quais",
      "quake",
      "quaky",
      "quale",
      "qualm",
      "quant",
      "quare",
      "quark",
      "quart",
      "quash",
      "quasi",
      "quass",
      "quate",
      "quay",
      "quays",
      "quean",
      "queen",
      "queer",
      "quell",
      "quern",
      "query",
      "quest",
      "queue",
      "quey",
      "queys",
      "quids",
      "quiet",
      "quiff",
      "quill",
      "quilt",
      "quins",
      "quint",
      "quips",
      "quipu",
      "quire",
      "quirt",
      "quite",
      "quits",
      "quod",
      "quods",
      "quoin",
      "quoit",
      "quota",
      "quote",
      "quoth",
      "qursh",
    ],
    "word": "quik",
  },
  {
    "columnNumber": 20,
    "context": "brown fox jumpd over the",
    "lineNumber": 1,
    "suggestions": [
      "jump",
      "jumps",
      "jumpy",
      "jumped",
      "jumbal",
      "jumble",
      "jumbo",
      "jumbos",
      "jumper",
    ],
    "word": "jumpd",
  },
  {
    "columnNumber": 51,
    "context": "near the beautifull lake Africa.",
    "lineNumber": 1,
    "suggestions": [
      "beautiful",
      "beautifuler",
      "beautifully",
    ],
    "word": "beautifull",
  },
  {
    "columnNumber": 91,
    "context": "as the briliant sun shining",
    "lineNumber": 1,
    "suggestions": [
      "brilliant",
    ],
    "word": "briliant",
  },
  {
    "columnNumber": 135,
    "context": "this fox performes incredible leeps,",
    "lineNumber": 1,
    "suggestions": [
      "performs",
      "performed",
      "performer",
      "performers",
      "perform",
      "performing",
    ],
    "word": "performes",
  },
  {
    "columnNumber": 151,
    "context": "performes incredible leeps, amuzing everyone",
    "lineNumber": 1,
    "suggestions": [
      "lees",
      "beeps",
      "deeps",
      "jeeps",
      "keeps",
      "neeps",
      "peeps",
      "seeps",
      "veeps",
      "weeps",
      "leaps",
      "leeks",
      "leers",
      "leets",
      "bleeps",
      "sleeps",
      "lee",
      "leech",
      "leek",
      "leer",
      "leered",
      "leery",
      "leet",
      "leeway",
    ],
    "word": "leeps",
  },
  {
    "columnNumber": 158,
    "context": "incredible leeps, amuzing everyone who",
    "lineNumber": 1,
    "suggestions": [
      "amazing",
      "amusing",
    ],
    "word": "amuzing",
  },
  {
    "columnNumber": 181,
    "context": "sees it. Truy, it is",
    "lineNumber": 1,
    "suggestions": [
      "try",
      "tray",
      "trey",
      "troy",
      "true",
      "trug",
      "truly",
      "trace",
      "track",
      "tract",
      "trad",
      "trade",
      "tragi",
      "traik",
      "trail",
      "train",
      "trait",
      "tram",
      "tramp",
      "trams",
      "trank",
      "tranq",
      "trans",
      "trap",
      "traps",
      "trapt",
      "trash",
      "trass",
      "trave",
      "trawl",
      "trays",
      "tread",
      "treat",
      "tree",
      "treed",
      "treen",
      "trees",
      "tref",
      "trek",
      "treks",
      "trend",
      "tress",
      "tret",
      "trets",
      "trews",
      "treys",
      "triac",
      "triad",
      "trial",
      "tribe",
      "trice",
      "trick",
      "tried",
      "trier",
      "tries",
      "trig",
      "trigo",
      "trigs",
      "trike",
      "trill",
      "trim",
      "trims",
      "trine",
      "trio",
      "triol",
      "trios",
      "trip",
      "tripe",
      "trips",
      "trite",
      "troak",
      "trock",
      "trod",
      "trode",
      "trois",
      "troke",
      "troll",
      "tromp",
      "trona",
      "trone",
      "troop",
      "trooz",
      "trop",
      "trope",
      "trot",
      "troth",
      "trots",
      "trout",
      "trove",
      "trow",
      "trows",
      "troys",
      "truce",
      "truck",
      "trued",
      "truer",
      "trues",
      "trugs",
      "trull",
      "trump",
      "trunk",
      "truss",
      "trust",
      "truth",
      "tryma",
      "tryst",
    ],
    "word": "Truy",
  },
  {
    "columnNumber": 197,
    "context": "is a magnifisent sight to",
    "lineNumber": 1,
    "suggestions": [
      "magnificent",
    ],
    "word": "magnifisent",
  },
]
`;
